### PR TITLE
Modifies query api for resources

### PR DESCRIPTION
**Pull request by PuneetPunamiya**
# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: api/design/resource.go [hunk 1]
```diff
@@ -30,13 +30,14 @@ var _ = Service("resource", func() {
 			Attribute("name", String, "Name of resource", func() {
 				Default("")
 			})
-			Attribute("kind", String, "Kind of resource", func() {
-				Enum("task", "pipeline", "")
-				Default("")
-			})
+			Attribute("kinds", ArrayOf(String), "Type of resource")
```
Comment (SM43) [inline api/design/resource.go:33]: could you update the `Description` for `kind`,`tags`,`exact` ?

File: api/pkg/service/resource/resource.go [hunk 2]
```diff
@@ -346,6 +360,37 @@ func filterByKind(t string) func(db *gorm.DB) *gorm.DB {
...
+}
+
+func filterByTags(t []string) func(db *gorm.DB) *gorm.DB {
+	if t == nil {
+		return noop
+	}
+
+	return func(db *gorm.DB) *gorm.DB {
+		t = lower(t)
+		query := db.Model(&model.Resource{}).Select("DISTINCT(resources.id)").
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:381]: move this to separate func?

File: api/pkg/service/resource/resource_test.go [hunk 3]
```diff
@@ -29,7 +29,7 @@ func TestQuery_DefaultLimit(t *testing.T) {
 	testutils.LoadFixtures(t, tc.FixturePath())
```
Comment (SM43) [inline api/pkg/service/resource/resource_test.go:29]: could you please also add HTTP test?

File: api/pkg/service/resource/resource.go [hunk 4]
```diff
@@ -346,6 +347,43 @@ func filterByKind(t string) func(db *gorm.DB) *gorm.DB {
...
+
+func filterByTags(t []string) func(db *gorm.DB) *gorm.DB {
+	if t == nil {
+		return noop
+	}
+
+	return func(db *gorm.DB) *gorm.DB {
+		t = lower(t)
+
+		resID := getResourceIDs(t, db)
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:369]: maybe we should call this from `Query` if tags != nil then call

Comment (SM43) [reply api/pkg/service/resource/resource.go:369]: for `filterByTags` only pass array of IDs and return the scope

File: api/pkg/service/resource/resource.go [hunk 5]
```diff
@@ -346,6 +347,43 @@ func filterByKind(t string) func(db *gorm.DB) *gorm.DB {
...
+
+		return db.Where("id IN (?)", resID)
+	}
+}
+
+func getResourceIDs(t []string, db *gorm.DB) (resID []uint) {
+
+	query := db.Model(&model.Resource{}).Select("DISTINCT(resources.id)").
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:377]: Is `Select("DISTINCT(resources.id)")` requires now if we are using pluck ?

File: api/pkg/service/resource/resource.go [hunk 6]
```diff
@@ -380,6 +418,25 @@ func matchesName(name string) func(db *gorm.DB) *gorm.DB {
...
+	if name == "" {
+		return noop
+	}
+	if exact {
+		return filterByName(name)
+	}
+	return matchesName(name)
+}
+
+// This function lowercase the first letter
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:431]: lowercase the first letter? or the whole string?

File: api/pkg/service/resource/resource_http_test.go [hunk 7]
```diff
@@ -42,7 +42,24 @@ func TestQuery_Http(t *testing.T) {
 	tc := testutils.Setup(t)
 	testutils.LoadFixtures(t, tc.FixturePath())
 
-	QueryChecker(tc).Test(t, http.MethodGet, "/query?name=build&kind=pipeline&limit=1").Check().
+	QueryChecker(tc).Test(t, http.MethodGet, "/query?name=build&kinds=pipeline&limit=1").Check().
```
Comment (SM43) [inline api/pkg/service/resource/resource_http_test.go:45]: could you please add more tests?
- passing only tags
- pass only kinds
- try to cover each case

File: api/pkg/service/resource/resource.go (hunk 5 above)
Comment (PuneetPunamiya) [reply api/pkg/service/resource/resource.go:377]: Using pluck it selects a single column from the table it does not select unique id's from the table

File: api/pkg/service/resource/resource.go (hunk 6 above)
Comment (PuneetPunamiya) [reply api/pkg/service/resource/resource.go:431]: It lowers the whole string, will update the comment

File: api/pkg/service/resource/resource.go [hunk 8]
```diff
@@ -48,10 +48,16 @@ func New(api app.BaseConfig) resource.Service {
 // Find resources based on name, kind or both
 func (s *service) Query(ctx context.Context, p *resource.QueryPayload) (res resource.ResourceCollection, err error) {
 
+	var resID []uint
+	if p.Tags != nil {
+		resID = getResourceIDs(p.Tags, s.db)
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:53]: queryResourcesByTags ?

File: api/pkg/service/resource/resource.go (hunk 5 above)
Comment (SM43) [reply api/pkg/service/resource/resource.go:380]: `Pluck("DISTINCT(resources.id)", &resID)` will this work?

Comment (PuneetPunamiya) [reply api/pkg/service/resource/resource.go]: Yes this will work !

File: api/pkg/service/resource/resource.go [hunk 9]
```diff
@@ -346,6 +352,41 @@ func filterByKind(t string) func(db *gorm.DB) *gorm.DB {
...
+
+func queryResourcesByTags(t []string, db *gorm.DB) (resID []uint) {
+	t = lower(t)
+
+	query := db.Model(&model.Resource{}).Select("resources.id").
+		Joins("JOIN resource_tags on resource_tags.resource_id = resources.id").
+		Joins("JOIN tags on tags.id = resource_tags.tag_id").
+		Where("tags.name in (?)", t)
+
+	query.Pluck("DISTINCT(resources.id)", &resID)
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:385]: Can you handle error in this query?

File: api/pkg/service/resource/resource.go [hunk 10]
```diff
@@ -346,6 +352,41 @@ func filterByKind(t string) func(db *gorm.DB) *gorm.DB {
...
+	return func(db *gorm.DB) *gorm.DB {
+
+		return db.Where("id IN (?)", t)
+	}
+}
+
+func queryResourcesByTags(t []string, db *gorm.DB) (resID []uint) {
+	t = lower(t)
+
+	query := db.Model(&model.Resource{}).Select("resources.id").
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:380]: do we need `Select("resources.id")` now?

File: api/pkg/service/resource/resource.go [hunk 11]
```diff
@@ -380,6 +421,24 @@ func matchesName(name string) func(db *gorm.DB) *gorm.DB {
...
+		return noop
+	}
+	if exact {
+		return filterByName(name)
+	}
+	return matchesName(name)
+}
+
+// This function lowercase all the elements of an array
+func lower(t []string) []string {
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:435]: can you put this below `noop`?

File: api/pkg/service/resource/resource.go [hunk 12]
```diff
@@ -346,6 +352,41 @@ func filterByKind(t string) func(db *gorm.DB) *gorm.DB {
...
+		return noop
+	}
+
+	return func(db *gorm.DB) *gorm.DB {
+
+		return db.Where("id IN (?)", t)
+	}
+}
+
+func queryResourcesByTags(t []string, db *gorm.DB) (resID []uint) {
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:377]: should we move this above `resourcesForQuery` and make this method of service and if the query fails return a `fetchError`?

**Comment by SM43** [comment]
@PuneetPunamiya maybe we should add support for `catalog` also as a query param? wdyt?

**Comment by PuneetPunamiya** [comment]
> @PuneetPunamiya maybe we should add support for `catalog` also as a query param? wdyt?

Yes I agree but thought once we have multi-catalog support then we can add it 
cc @sthaha

File: api/design/resource.go [hunk 13]
```diff
@@ -30,13 +30,14 @@ var _ = Service("resource", func() {
...
-			Attribute("kind", String, "Kind of resource", func() {
-				Enum("task", "pipeline", "")
-				Default("")
-			})
+			Attribute("kinds", ArrayOf(String), "Kinds of resource")
+			Attribute("tags", ArrayOf(String), "Tags associated with a resource")
 			Attribute("limit", UInt, "Maximum number of resources to be returned", func() {
 				Default(100)
 			})
+			Attribute("exact", Boolean, "Find the exact resource", func() {
```
Comment (SM43) [inline api/design/resource.go:38]: you can use an `Enum` here I guess to restrict the input...`true` and `false`

Comment (PuneetPunamiya) [reply api/design/resource.go:38]: Since it is boolean do we still need to use enum ?

Comment (SM43) [reply api/design/resource.go:38]: if we pass anything other that, does it return an error?

Comment (SM43) [reply api/design/resource.go:38]: then if is fine

Comment (PuneetPunamiya) [reply api/design/resource.go:38]: Yes that would return an error

Comment (PuneetPunamiya) [reply api/design/resource.go:38]: Saying the type of exact should be boolean

File: api/pkg/service/resource/resource.go [hunk 14]
```diff
@@ -158,6 +166,23 @@ func (s *service) ByID(ctx context.Context, p *resource.ByIDPayload) (res *resou
 	return res, nil
 }
 
+func (s *service) queryResourceIDByTags(t []string) ([]uint, error) {
+	t = lower(t)
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:170]: should we rename `lower` to something else? :thinking: this is kinda odd

File: api/pkg/service/resource/resource.go [hunk 15]
```diff
@@ -346,6 +371,28 @@ func filterByKind(t string) func(db *gorm.DB) *gorm.DB {
...
+		return noop
+	}
+
+	return func(db *gorm.DB) *gorm.DB {
+		t = lower(t)
+		return db.Where("LOWER(kind) IN (?)", t)
+	}
+}
+
+func filterByTags(t []uint) func(db *gorm.DB) *gorm.DB {
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:385]: this is actually filter by id, we are passing resource ids... should we rename it to `filterByIds`

File: api/pkg/service/resource/resource.go [hunk 16]
```diff
@@ -48,10 +48,18 @@ func New(api app.BaseConfig) resource.Service {
 // Find resources based on name, kind or both
 func (s *service) Query(ctx context.Context, p *resource.QueryPayload) (res resource.ResourceCollection, err error) {
 
+	var resID []uint
+	if p.Tags != nil {
+		if resID, err = s.queryResourceIDByTags(p.Tags); err != nil {
```
Comment (SM43) [inline api/pkg/service/resource/resource.go:53]: this can be filterByTags :D

**Review by sthaha — COMMENTED** [review]
/lgtm
/approve
/meow

**Pull request merged**